File: LeetCode/0053. Maximum Subarray/solution.py
"""
209 / 209 test cases passed.
Runtime: 248 ms
Memory Usage: 26 MB
"""
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        dp = [0] * (len(nums) + 1)
        ans = nums[0]
        for i, num in enumerate(nums):
            dp[i + 1] = max(dp[i] + num, num)
            ans = max(ans, dp[i + 1])
        return ans


"""
209 / 209 test cases passed.
Runtime: 180 ms
Memory Usage: 25.7 MB
"""
class Solution2:
    def maxSubArray(self, nums: List[int]) -> int:
        ans, prev = nums[0], 0
        for i, num in enumerate(nums):
            curr = max(prev + num, num)
            ans = max(ans, curr)
            prev = curr
        return ans
        

File: LeetCode/0053. Maximum Subarray/Solution.java
/**
 * 209 / 209 test cases passed.
 * Runtime: 1 ms
 * Memory Usage: 50.4 MB 
 */
class Solution {
    public int maxSubArray(int[] nums) {
        int ans = nums[0], prev = 0;
        for (int i = 0; i < nums.length; ++ i) {
            int curr = Math.max(nums[i], prev + nums[i]);
            ans = Math.max(ans, curr);
            prev = curr;
        }
        return ans;
    }
}
